contrib/fuzz/metadata_fuzzer.go

```diff
@@ -0,0 +1,406 @@
+   limitations under the License.
+*/
+
+
+package fuzz
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	fuzz "github.com/AdaLogics/go-fuzz-headers"
+	digest "github.com/opencontainers/go-digest"
+	ocispec "github.com/opencontainers/image-spec/specs-go/v1"
+	bolt "go.etcd.io/bbolt"
+
+	"github.com/containerd/containerd/containers"
+	"github.com/containerd/containerd/content"
+	"github.com/containerd/containerd/content/local"
+	"github.com/containerd/containerd/images"
+	"github.com/containerd/containerd/leases"
+	"github.com/containerd/containerd/metadata"
+	"github.com/containerd/containerd/namespaces"
+	"github.com/containerd/containerd/snapshots"
+	"github.com/containerd/containerd/snapshots/native"
+)
+
+func testEnv() (context.Context, *bolt.DB, func(), error) {
+	ctx, cancel := context.WithCancel(context.Background())
+	ctx = namespaces.WithNamespace(ctx, "testing")
+
+	dirname, err := ioutil.TempDir("", "fuzz-")
+	if err != nil {
+		return ctx, nil, func() { cancel() }, err
+	}
+
+	db, err := bolt.Open(filepath.Join(dirname, "meta.db"), 0644, nil)
+	if err != nil {
+		return ctx, nil, func() { cancel() }, err
+	}
+
+	return ctx, db, func() {
+		db.Close()
+		_ = os.RemoveAll(dirname)
+		cancel()
+	}, nil
+}
+
+func FuzzImageStore(data []byte) int {
+	ctx, db, cancel, err := testEnv()
+	if err != nil {
+		return 0
+	}
+	defer cancel()
+	store := metadata.NewImageStore(metadata.NewDB(db, nil, nil))
+	f := fuzz.NewConsumer(data)
+	noOfOperations, err := f.GetInt()
+	if err != nil {
+		return 0
+	}
+	maxOperations := 50
+	for i := 0; i < noOfOperations%maxOperations; i++ {
+		opType, err := f.GetInt()
+		if err != nil {
+			return 0
+		}
+		if opType%1 == 0 {
+			i := images.Image{}
+			err := f.GenerateStruct(&i)
+			if err != nil {
+				return 0
+			}
+			_, _ = store.Create(ctx, i)
+		} else if opType%2 == 0 {
+			newFs, err := f.GetString()
+			if err != nil {
+				return 0
+			}
+			_, _ = store.List(ctx, newFs)
+		} else if opType%3 == 0 {
+			i := images.Image{}
+			err := f.GenerateStruct(&i)
+			if err != nil {
+				return 0
+			}
+			_, _ = store.Update(ctx, i)
+		} else if opType%4 == 0 {
+			name, err := f.GetString()
+			if err != nil {
+				return 0
+			}
+			_ = store.Delete(ctx, name)
+		}
+	}
+	return 1
+}
+
+func FuzzLeaseManager(data []byte) int {
+	ctx, db, cancel, err := testEnv()
+	if err != nil {
+		return 0
+	}
+	defer cancel()
+	lm := metadata.NewLeaseManager(metadata.NewDB(db, nil, nil))
+
+	f := fuzz.NewConsumer(data)
+	noOfOperations, err := f.GetInt()
+	if err != nil {
+		return 0
+	}
+	maxOperations := 50
+	for i := 0; i < noOfOperations%maxOperations; i++ {
+		opType, err := f.GetInt()
+		if err != nil {
+			return 0
+		}
+		if opType%1 == 0 {
+			err := db.Update(func(tx *bolt.Tx) error {
+				sm := make(map[string]string)
+				err2 := f.FuzzMap(&sm)
+				if err2 != nil {
+					return err2
+				}
+				_, _ = lm.Create(ctx, leases.WithLabels(sm))
```

Comment:
Don't you need to check the error here?

Reply:
No. The logic is that if the fuzzer fails to create a new lease, the lease manager should still be functional, and the fuzzer tests if that is not the case.